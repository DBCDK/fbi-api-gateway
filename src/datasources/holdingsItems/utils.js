// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
