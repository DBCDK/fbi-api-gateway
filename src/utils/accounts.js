/**
 * @file This file handles util function to select single or multiple CULR accounts
 *
 */

import { isValidCpr } from "./cpr";

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} Object or null
 */
export async function getAccount(accessToken, context, props) {
  return (await getAccounts(accessToken, context, props))?.[0] || null;
}

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} (Array)
 */
export async function getAccounts(accessToken, context, props) {
  // userInfo
  const userinfo = await context.datasources.getLoader("userinfo").load({
    accessToken,
  });

  const user = userinfo?.attributes;

  if (!user?.userId) {
    return [];
  }

  // select dataloader
  let dataloader = isValidCpr(user.userId)
    ? "culrGetAccountsByGlobalId"
    : "culrGetAccountsByLocalId";

  // Retrieve user culr account
  const response = await context.datasources.getLoader(dataloader).load({
    userId: user.userId,
    agencyId: user.loggedInAgencyId,
  });

  return filterAccountsByProps(response.accounts, props);
}

/**
 *
 * @param {Array} accounts
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAccountsByProps(accounts = [], props = {}) {
  if (props.agency) {
    accounts = accounts?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    accounts = accounts?.filter(({ userIdValue }) => userIdValue === props.id);
  }
  if (props.type) {
    accounts = accounts?.filter(({ userIdType }) => userIdType === props.type);
  }
  return accounts;
}

/**
 *
 * Function to select a specific agency from a userinfo agencies list
 *
 * @param {Array} agencies
 * @param {object} props - Filter agencies by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAgenciesByProps(agencies = [], props = {}) {
  if (props.agency) {
    agencies = agencies?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    agencies = agencies?.filter(({ userId }) => userId === props.id);
  }
  if (props.type) {
    agencies = agencies?.filter(({ userIdType }) => userIdType === props.type);
  }
  return agencies;
}
