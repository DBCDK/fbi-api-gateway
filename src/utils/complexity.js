import { log } from "dbc-node-logger";
import { parse, GraphQLError } from "graphql";

import queryComplexity, {
  getComplexity,
  simpleEstimator,
  directiveEstimator,
} from "graphql-query-complexity";

import config from "../config";

/**
 * Function to build the content for the estimator functions
 *
 * @param {*} params.query
 * @param {*} params.variables
 * @param {object} params.schema
 * @returns {object}
 */
export function buildQueryComplexity({
  schema = null,
  query,
  variables,
  maxQueryComplexity = null,
}) {
  // Set the complexity limit (custom or default/config)
  const COMPLEXITY_LIMIT = maxQueryComplexity || config.query.maxComplexity;

  return {
    estimators: [
      directiveEstimator(),
      simpleEstimator({
        defaultComplexity: 1,
      }),
    ],
    schema,
    maximumComplexity: COMPLEXITY_LIMIT,
    query: parse(query),
    variables,
    createError: (max, actual) => {
      return new GraphQLError(
        `Query is too complex: ${actual}. Maximum allowed complexity: ${max}`
      );
    },
    onComplete: (complexity) => {
      if (complexity > COMPLEXITY_LIMIT) {
        log.error("Query exceeded complexity limit", {
          complexity,
          query,
          clientConfiguredLimit: !!maxQueryComplexity,
          maxComplexity: COMPLEXITY_LIMIT,
        });
      }
    },
  };
}

/**
 * Future complexity estimator (returns validation function)
 *
 * @param {*} params.query
 * @param {*} params.variables
 * @param {object} params.schema
 * @returns {funciton}
 */

export function validateQueryComplexity(params) {
  try {
    return queryComplexity(buildQueryComplexity(params));
  } catch (e) {
    // Log error in case complexity cannot be calculated (invalid query, misconfiguration, etc.)
  }
}

/**
 * Future complexity estimator (returns complexity value)
 *
 * @param {*} params.query
 * @param {*} params.variables
 * @param {object} params.schema
 * @returns {int}
 */

export function getQueryComplexity(params) {
  try {
    return getComplexity(buildQueryComplexity(params));
  } catch (e) {
    // Log error in case complexity cannot be calculated (invalid query, misconfiguration, etc.)
  }
}

/**
 *
 * @param {int} complexity
 * @returns {string}
 */
export function getQueryComplexityClass(complexity) {
  if (complexity < 250) {
    return "simple";
  }
  if (complexity < 2500) {
    return "complex";
  }
  if (complexity < 25000) {
    return "critical";
  }

  return "rejected";
}
