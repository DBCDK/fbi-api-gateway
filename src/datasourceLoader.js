import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
