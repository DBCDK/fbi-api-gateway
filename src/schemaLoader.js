/**
 * @file Responsible for loading the schema
 *
 * It will check all files recursively in ./src/schema
 * and load type definitions and resolvers
 */
import { startCase } from "lodash";

import { makeExecutableSchema, mergeSchemas } from "@graphql-tools/schema";
import { wrapSchema } from "@graphql-tools/wrap";
import { mergeTypeDefs } from "@graphql-tools/merge";
import { filterSchema, pruneSchema } from "@graphql-tools/utils";

import { typeDefs as scalarTypeDefs } from "graphql-scalars";
import { resolvers as scalarResolvers } from "graphql-scalars";
import { log } from "dbc-node-logger";

import drupalSchema from "./schema/external/drupal";
import { getFilesRecursive } from "./utils/utils";
import { wrapResolvers } from "./utils/wrapResolvers";

import merge from "lodash/merge";
import { parseClientPermissions } from "../commonUtils";
import enumFallbackDirective from "./utils/enumFallbackDirective";

const { enumFallbackDirectiveTypeDefs, enumFallbackDirectiveTransformer } =
  enumFallbackDirective();

// Stores the transformed schemas
const schemaCache = {};

// The external schema (headless Drupal)
let externalSchema;

// The internal schema
let internalSchema = enumFallbackDirectiveTransformer(
  makeExecutableSchema(fieldNameValidatorEnv(schemaLoader()))
);

/**
 * PermissionTransform  is used to remove parts of the schema
 * as dictated by a given smaug client configuration
 */
class PermissionsTransform {
  constructor(clientPermissions) {
    this.clientPermissions = clientPermissions;
  }
  transformSchema(originalWrappingSchema) {
    return pruneSchema(
      filterSchema({
        schema: originalWrappingSchema,
        rootFieldFilter: (operationName, fieldName) =>
          this.clientPermissions?.allowRootFields?.includes(fieldName),
        typeFilter: (typeName) =>
          !this.clientPermissions?.denyTypes?.includes(typeName),
        // fieldFilter: (typeName, fieldName) => true,
        // argumentFilter: (typeName, fieldName, argName) => true
      })
    );
  }
}

/**
 * Runs the fieldNameValidator function if environment is NOT production
 *
 * @param {Schema} props
 * @param {string} errorType (THROW | LOG | IGNORE)
 * @returns {object}
 */
function fieldNameValidatorEnv(props, errorType = "THROW") {
  if (process.env.NODE_ENV === "production") {
    return props;
  }

  return fieldNameValidator(props, errorType);
}

/**
 * Type, field/subfield validator
 *
 * Function is testing for:
 *
 * ObjectTypes i written in PascalCase (starting with a Capital letter)
 * Unique Type subfield names (regardless of UPPER/lower case)
 * EnumType values is written in all UPPERCASE
 * Types written in PascalCase
 * InputTypes endping with 'Input'
 * ScalarTypes ending with 'Scalar'
 * EnumTypes ending with 'Enum'
 * UnionTypes ending with 'Union'
 * InterfaceTypes ending with 'Interface'
 *
 * Notes: The naming standard for deprecated fields are ignored ---> this check can be removed in future
 * Directive definitions are excepted
 *
 * Deprecation 'reason' are also validated.
 * A deprecation reason should contain a expration date in the dd/mm-yyyy format.
 *
 * @param {Schema} props
 * @param {string} errorType (THROW | LOG | IGNORE)
 * @returns {object}
 */

// dev mode only
export function fieldNameValidator(props, errorType = "THROW") {
  //  This check is NOT for production

  if (errorType === "LOG") {
    console.info(
      "########################### Field validations ###########################"
    );
    console.info(" ");
  }

  let hasError;

  // Function to handle Error messages (throw|log|ignore)
  function handleError(message) {
    hasError = true;
    if (errorType === "THROW") {
      throw new Error(message);
    }
    errorType === "LOG" && console.error(message);
  }

  // Kind/type categories to omit
  const omit = ["DirectiveDefinition"];

  // Kinds/types which should be a part of the tailing typeNameDefintion. e.g. UserStatus(Enum), User(Input)
  const tailedTypes = [
    "EnumTypeDefinition",
    "InputObjectTypeDefinition",
    "UnionTypeDefinition",
    "ScalarTypeDefinition",
    "InterfaceTypeDefinition",
  ];

  // All fields and subfields in lowercase
  let lowerCasedTypeNames = [];

  // schema
  const typeDefs = props.typeDefs;

  typeDefs?.definitions?.forEach((obj) => {
    const field = obj?.name?.value;
    const shouldOmit = omit.includes(obj?.kind);
    const shouldHaveTypedTailing = tailedTypes.includes(obj?.kind);

    // field name exist
    if (field) {
      // Tjek if the type of field is omitted
      if (!shouldOmit) {
        // Check for field name doublets
        if (lowerCasedTypeNames.includes(field?.toLowerCase())) {
          handleError(`Type name '${field}' (${obj?.kind}) is already used`);
        }

        // Add to name doublet array
        lowerCasedTypeNames.push(field?.toLowerCase());

        // Ensure first letter in type def is uppercase
        if (/^\p{Ll}/u.test(field)) {
          handleError(
            `Type '${field}' (${obj?.kind}) is a TypeDefinition and should be written in PascalCase`
          );
        }

        // Ensure correct tailed naming for Enum and Input types
        if (shouldHaveTypedTailing) {
          // Returns the first string in PascalCased type/kind e.g. Enum, Union, Interface
          const tail = startCase(obj?.kind).split(" ")[0];

          if (!field.endsWith(tail)) {
            handleError(
              `Type '${field}' is a ${obj?.kind} which should always end with '${tail}'`
            );
          }
        }
      }

      /**
       * Subfields check
       */

      // ObjectTypeDefinition uses .fields and EnumValueDefinition uses .values
      const subfields = obj?.fields || obj?.values;

      // All subfields in lowercase
      let lowerCasedFieldNames = [];

      // Check the subfields of the typedef obj
      subfields?.forEach((obj) => {
        const subfield = obj?.name?.value;
        const kind = obj?.kind;
        const isDeprecated = !!obj?.directives?.find(
          (obj) => obj?.name?.value === "deprecated"
        );

        // ignore deprecated fields
        if (!isDeprecated) {
          // has subfields
          if (subfield) {
            // handle enum subfields (all UPPERCASE check)
            if (kind === "EnumValueDefinition") {
              //  ensure enums is written all uppercase
              if (subfield !== subfield?.toUpperCase()) {
                handleError(
                  `Field '${subfield}' in type '${field}' (${obj?.kind}) is an enum and should be written all UPPERCASE`
                );
              }
            }

            // handle if subfield name already exist
            if (lowerCasedFieldNames.includes(subfield?.toLowerCase())) {
              handleError(
                `Field name '${subfield}' in '${field}' is already used`
              );
            }

            //  add to name doublet array
            lowerCasedFieldNames.push(subfield?.toLowerCase());
          }
        }
      });

      /**
       * Deprecated fields check
       */
      subfields?.forEach((obj) => {
        const subfield = obj?.name?.value;
        const isDeprecated = !!obj?.directives?.find(
          (obj) => obj?.name?.value === "deprecated"
        );

        if (isDeprecated) {
          const target = obj?.directives?.find(
            (obj) => obj?.name?.value === "deprecated"
          );

          const reason = target?.arguments?.find(
            ({ name }) => name?.value === "reason"
          );

          const value = reason?.value?.value;

          // ensure reason includes an expires string and date has correct format
          if (
            !/expires: (0[1-9]|[12][0-9]|3[01])\/(0[1-9]|1[1,2])-(20)\d{2}/.test(
              value
            )
          ) {
            handleError(
              `Deprecated field name '${subfield}' in '${field}' has wrong deprecation format. 

              >>> Deprecation should contain a short text with an 'expires: dd/mm-yyyy suffix'
              `
            );
          }
        }
      });
    }
  });

  //  If error handling type is "LOG"
  if (errorType === "LOG") {
    if (!hasError) {
      console.info("... No field validation errors was found");
    }
    console.info(" ");
    console.info(
      "#########################################################################"
    );
  }

  return props;
}

/**
 * Will load all files in schema folder
 * and look for type definitions and resolvers.
 */
export function schemaLoader() {
  // Custom selected scalar type defs (from graphiql-scalar lib)
  const customScalarTypeDefs = ["DateTime"];

  const _scalarResolvers = {};
  const _scalarTypeDefs = [];

  customScalarTypeDefs.forEach((val) => {
    if (scalarTypeDefs.includes(`scalar ${val}`)) {
      _scalarResolvers[`${val}Scalar`] = scalarResolvers[val];
      _scalarTypeDefs.push(`scalar ${val}Scalar`);
    }
  });

  let allTypeDefs = [enumFallbackDirectiveTypeDefs, ..._scalarTypeDefs];

  let allResolvers = { ..._scalarResolvers };

  // Load files in schema folder
  const files = getFilesRecursive(`${__dirname}/schema`);

  // Require typeDefs and resolvers
  files.forEach((file) => {
    if (!file.path.endsWith(".js")) {
      return;
    }
    const { typeDef, resolvers } = require(file.path);
    if (typeDef) {
      allTypeDefs = [...allTypeDefs, typeDef];
      log.debug(`Found type definition in ${file.path}`);
    }
    if (resolvers) {
      allResolvers = merge({}, allResolvers, resolvers);
    }
  });

  return { typeDefs: mergeTypeDefs(allTypeDefs), resolvers: allResolvers };
}

/**
 * Gets an executable schema that is transformed
 * according to the permissions of the smaug client.
 */
export async function getExecutableSchema({
  loadExternal = true,
  clientPermissions,
  hasAccessToken,
}) {
  const parsedPermissions = parseClientPermissions({
    smaug: clientPermissions,
  });

  const key = JSON.stringify({ hasAccessToken, parsedPermissions });

  if (!schemaCache[key]) {
    // Fetch external Drupal schema (bibdk)
    if (!externalSchema && loadExternal) {
      externalSchema = await drupalSchema();
    }

    // Merge external and internal schemas
    const mergedSchema = loadExternal
      ? mergeSchemas({
          schemas: [externalSchema, internalSchema],
        })
      : internalSchema;

    // Filter schema according to permissions of Smaug client
    // If no valid accessToken is given (no smaug client), we allow to introspect
    // the entire schema (useful for test purposes). But no data is accessible.
    const filteredSchema =
      parsedPermissions?.admin || !hasAccessToken
        ? mergedSchema
        : wrapSchema({
            schema: mergedSchema,
            transforms: [new PermissionsTransform(parsedPermissions)],
          });

    // Wrap all resolvers with error logger
    wrapResolvers(filteredSchema, (resolveFn) => {
      async function errorLogger(...args) {
        const result = await resolveFn(...args);
        if (result instanceof Error) {
          log.error(result.message, {
            error: String(result),
            stacktrace: result.stack,
          });
        }
        return result;
      }
      return errorLogger;
    });

    schemaCache[key] = filteredSchema;
  }

  return schemaCache[key];
}
