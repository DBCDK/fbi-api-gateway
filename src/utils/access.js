/**
 * @file access util functions
 */

import { filterDuplicateAgencies, resolveManifestation } from "./utils";

/**
 *
 * Check whether any of the users libraries has access to Infomedia.
 * The returned agencyId will be billed for infomedia access.
 *
 * First we check if the users login library has access to info media, if not,
 * we check if any other og the users associated accounts has access to infomedia,
 * if so, we return the first one found.
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
async function infomedia(context) {
  const user = context?.user;

  // get rights from idp
  const idpRights = await context.datasources.getLoader("idp").load("");

  // check if users loggedInAgency has infomedia access
  const loggedInAgencyId = user?.loggedInAgencyId;

  if (loggedInAgencyId && idpRights[loggedInAgencyId]) {
    return loggedInAgencyId;
  }

  /*
   * Alternativly check all users accounts
   */

  // filtered accounts
  const userInfoAccounts = filterDuplicateAgencies(user.agencies);

  const hasAccess = userInfoAccounts?.filter(
    ({ agencyId }) => idpRights[agencyId]
  );

  // check for infomedia access - if any of users agencies subscribes
  if (hasAccess.length > 0) {
    return hasAccess[0]?.agencyId;
  }

  return null;
}

/**
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
export async function getInfomediaAgencyId(context) {
  return await infomedia(context);
}

/**
 *
 * @param {object} context
 *
 * @returns {boolean}
 */
export async function hasInfomediaAccess(context) {
  return !!(await infomedia(context));
}

/**
 * Get type of access for manifestation with given pid.
 *  url access
 *  ill access
 *  infomedia access
 *  digital article service access
 *
 * @param pid
 * @param context
 * @returns {Promise<*>}
 */
export async function resolveAccess(manifestation, context) {
  // We parse the access structure from JED, and convert it
  // to the union type structure.
  // At some point we may choose to follow the structure of JED closely,
  // but it has to be coordinated with stakeholders

  let parent =
    typeof manifestation === "object"
      ? manifestation
      : await resolveManifestation({ pid: manifestation }, context);

  const res = [];

  // use linkchecker to check status of a single accessUrl
  // linkchecker does NOT handle proxy urls, so we skip ebookcentral and ebsco urls
  const linkStatus = async (url) => {
    if (url?.includes("ebookcentral") || url?.includes("ebscohost")) {
      return "OK";
    }

    const linkcheck = await context.datasources
      .getLoader("linkcheck")
      .load({ urls: [url] });
    const link = linkcheck.find((check) => check.url === url);

    return link?.status || "OK";
  };

  parent?.access?.accessUrls?.forEach((entry) => {
    const { proxyUrl, loginRequired } = getProxyUrl(
      entry.url || "",
      context?.user
    );

    res.push({
      __typename: "AccessUrl",
      origin: parseOnlineUrlToOrigin(entry.url),
      url: proxyUrl,
      loginRequired,
      note: entry.note,
      type: entry.type,
      status: linkStatus(proxyUrl) || "OK",
    });
  });

  if (parent?.access?.dbcWebArchive) {
    const archives = await context.datasources
      .getLoader("moreinfoWebarchive")
      .load(parent.pid);

    archives.forEach((archive) => {
      if (archive.url) {
        res.push({
          __typename: "AccessUrl",
          url: archive.url,
          origin: parseOnlineUrlToOrigin(archive.url),
          loginRequired: false,
          type: "RESOURCE",
          status: linkStatus(archive.url) || "OK",
        });
      }
    });
  }

  parent?.access?.ereol?.forEach((entry) => {
    res.push({
      __typename: "Ereol",
      origin: entry.origin,
      url: entry.url,
      canAlwaysBeLoaned: entry.canAlwaysBeLoaned,
      note: entry.note,
    });
  });

  if (parent?.access?.infomediaService?.id) {
    if (
      !["politiken", "jyllands-posten"].some((publication) =>
        parent?.hostPublication?.title?.toLowerCase()?.includes(publication)
      )
    ) {
      res.push({
        __typename: "InfomediaService",
        id: parent.access.infomediaService.id,
      });
    }

    // If this is a omitted online access for politiken or jyllands-posten.
    else {
      const hasPhysical = parent?.accessTypes?.some(
        (t) => (t?.code ?? "").toUpperCase() === "PHYSICAL"
      );

      if (hasPhysical) {
        parent = {
          ...parent,
          access: {
            ...(parent.access ?? {}),
            interLibraryLoanIsPossible: true,
          },
        };
      }
    }
  }

  // Get the issn for this article or periodica
  const issn =
    parent?.access?.digitalArticleService?.issn ||
    parent?.identifiers?.find?.((id) => id.type === "ISSN")?.value;

  if (issn) {
    const journals = await context.datasources
      .getLoader("statsbiblioteketJournals")
      .load("");
    const articleissn = issn.replace(/[^a-z\d]/gi, "");
    const hasJournal = journals && journals[articleissn];
    if (hasJournal) {
      res.push({
        __typename: "DigitalArticleService",
        issn: articleissn,
      });
    }
  }

  if (parent?.access?.interLibraryLoanIsPossible) {
    //if collectionidentifiers includes 870970-accessnew it may be loaned
    const isNewMaterial =
      parent?.collectionIdentifiers?.includes("870970-accessnew");

    const forLoan = checkInterLibraryLoan(parent, context);
    res.push({
      __typename: "InterLibraryLoan",
      loanIsPossible: forLoan,
      accessNew: isNewMaterial,
    });
  }

  // Return array containing all types of access
  return _sortOnlineAccess(res);
}

/**
 * municipality number is the second|third|fourth digit in agencyId
 * @param agencyId
 * @returns {string | undefined}
 */
function parseForMunicipalityNumber(agencyId) {
  return agencyId?.substring(1, 4);
}

/**
 * This one is for ebook.plus - we need to go via a proxy url (if user is logged in)
 * @param url
 * @param user
 * @returns {*}
 */
export function getProxyUrl(url, user) {
  const municipality =
    user?.municipality ||
    parseForMunicipalityNumber(user?.municipalityAgencyId);

  // check if we should proxy this url - for now it is ebookcentral and ebscohost
  const proxyMe =
    url.indexOf("ebookcentral") !== -1 || url.indexOf("ebscohost") !== -1;
  if (proxyMe) {
    // check if user is logged in
    if (user?.userId) {
      const realUrl = `https://bib${municipality}.bibbaser.dk/login?url=${url}`;
      return { proxyUrl: realUrl, loginRequired: proxyMe };
    }
  }

  return { proxyUrl: url, loginRequired: proxyMe };
}

/**
 * Some sources are not for loan (fagbibliografier) - for now a statis list:
 * This method is only called if interLibraryLoan is true - We know that for a fact
 * 159080-fagbib - Besættelsesbibliografien
 * 159081-fagbib - Bibliografi over Dansk Kunst
 * 159082-fagbib - Dansk Historisk Bibliografi
 * 159083-fagbib - Dansk Musiklitterær Bibliografi
 * 159084-fagbib - Kongelige Teater programartikler
 * 159085-fagbib - Dania Polyglotta
 * 159086-fagbib - Sportline
 * @param parent
 * @param context
 */
function checkInterLibraryLoan(parent, context) {
  // This is a blacklist of bases NOT to be loaned
  const notForLoanList = [
    "159080-fagbib",
    "159081-fagbib",
    "159082-fagbib",
    "159083-fagbib",
    "159084-fagbib",
    "159085-fagbib",
    "159086-fagbib",
  ];
  // we check on objectId - first part is the source
  const source = parent?.objectId?.split(":")[0];
  return !notForLoanList.includes(source);
}

/**
 * Get domain from url
 * @param url
 * @return String
 *  A parsed string eg. "DBC Webarkiv" or name of host eg. "infolink2003.elbo.dk"
 *
 * */
export function parseOnlineUrlToOrigin(url) {
  try {
    const parsedUrl = new URL(url);
    if (parsedUrl["host"] === "moreinfo.addi.dk") {
      return "DBC Webarkiv";
    } else {
      return (parsedUrl["host"] && parsedUrl["host"]) || "";
    }
  } catch (e) {
    return "";
  }
}

/**
 * Handle special cases - for now we want filmstriben/fjernleje BEFORE filmstriben/biblioteket AND
 * we want dansk film institut (dfi) to come last - it is not a 'real' url but a link to a description
 * @param onlineAccess
 * @return {*}
 * @private
 */
function _sortOnlineAccess(onlineAccess) {
  const specialSort = (a, b) => {
    // fjernleje should be on top
    if (b.url && b.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return 1;
    } else if (a.url && a.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return -1;
      // dfi is not a 'real' online url - sort low
    } else if (b.url && b.url.indexOf("dfi.dk") !== -1) {
      return -1;
    } else if (a.url && a.url.indexOf("dfi.dk") !== -1) {
      return 1;
    }
    return 0;
  };

  return onlineAccess.sort(specialSort);
}
