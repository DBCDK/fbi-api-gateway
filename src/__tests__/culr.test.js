/**
 * @file This file tests the complexity estimator function, which is evaluated in in graphql validator
 *
 */

import { getExecutableSchema } from "../schemaLoader";

import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";

let internalSchema;
describe("Culr", () => {
  beforeEach(async () => {
    if (!internalSchema) {
      internalSchema = await getExecutableSchema({
        loadExternal: false,
        clientPermissions: { admin: true },
      });
    }
  });

  it.only("Should give status ERROR_INVALID_CPR", async () => {
    const query = `
    mutation Example_CreateAccount($input: CreateAccountInput!) {
        culr {
            createAccount(input: $input, dryRun: true) {
                status
            }
        }
    }
    `;

    const result = await performTestQuery({
      query,
      variables: {
        input: {
          agencyId: "800010",
          localId: "C012345678",
          cpr: "0123456789",
        },
      },
      context: {
        datasources: createMockedDataLoaders(),
        accessToken: "DUMMY_TOKEN",
      },
    });

    console.error("ffffffff", result);

    expect(result).toMatchSnapshot();
  });
});
