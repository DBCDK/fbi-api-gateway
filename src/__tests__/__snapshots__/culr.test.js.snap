// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Culr Should give status ERROR_INVALID_CPR 1`] = `
Object {
  "data": Object {
    "culr": Object {
      "createAccount": Object {
        "status": "ERROR_UNAUTHENTICATED_TOKEN",
      },
    },
  },
}
`;
